File: ejercicio4/find_check.py
class FindCheck:
    def __init__(self, board):
        self.board = board

    def find_king(self, board, piece):
        for row_index, row in enumerate(board):
            col_index = row.find(piece)
            if col_index != -1:
                return row_index, col_index
        return None

    def piece_color(self, piece):
        return "white" if piece == piece.upper() else "black"

    def pawn_check(self, pawn, row_index, col_index, king):
        potential_checks = [
            (row_index - 1, col_index - 1),
            (row_index - 1, col_index + 1)
        ]
        return king in potential_checks

    def knight_check(self, piece, row_index, col_index, king):
        potential_checks = [
            (row_index - 2, col_index - 1),
            (row_index - 2, col_index + 1),
            (row_index - 1, col_index - 2),
            (row_index - 1, col_index + 2),
            (row_index + 1, col_index - 2),
            (row_index + 1, col_index + 2),
            (row_index + 2, col_index - 1),
            (row_index + 2, col_index + 1)
        ]
        return king in potential_checks

    def diagonal_check(self, piece, row_index, col_index, king):
        king_row, king_col = king
        if abs(row_index - king_row) != abs(col_index - king_col):
            return False
        x_dir = 1 if king_col > col_index else -1
        y_dir = 1 if king_row > row_index else -1
        x, y = col_index + x_dir, row_index + y_dir
        while x != king_col:
            if self.board[y][x] != "0":
                return False
            x += x_dir
            y += y_dir
        return True

    def vertical_check(self, piece, row_index, col_index, king):
        if king is None:
            return False
        king_row, king_col = king
        if col_index != king_col:
            return False
        y_dir = 1 if king_row > row_index else -1
        y = row_index + y_dir
        while y != king_row:
            if self.board[y][col_index] != "0":
                return False
            y += y_dir
        return True

    def horizontal_check(self, piece, row_index, col_index, king):
        if king is None:
            return False
        king_row, king_col = king
        if row_index != king_row:
            return False
        x_dir = 1 if king_col > col_index else -1
        x = col_index + x_dir
        while x != king_col:
            if self.board[row_index][x] != "0":
                return False
            x += x_dir
        return True

    def check(self):
        white_king = self.find_king(self.board, "R")
        black_king = self.find_king(self.board, "r")
        for row_index, row in enumerate(self.board):
            for col_index, piece in enumerate(row):
                king = black_king if self.piece_color(piece) == "white" else white_king
                if piece not in ["0", "R", "r"]:
                    if piece in ["P", "p"] and self.pawn_check(piece, row_index, col_index, king):
                        return True
                    elif piece in ["A", "a"] and self.diagonal_check(piece, row_index, col_index, king):
                        return True
                    elif piece in ["C", "c"] and self.knight_check(piece, row_index, col_index, king):
                        return True
                    elif piece in ["T", "t"] and (self.vertical_check(piece, row_index, col_index, king) or self.horizontal_check(piece, row_index, col_index, king)):
                        return True
                    elif piece in ["D", "d"] and (self.diagonal_check(piece, row_index, col_index, king) or self.vertical_check(piece, row_index, col_index, king) or self.horizontal_check(piece, row_index, col_index, king)):
                        return True
        return False

File: ejercicio4/old_ruby_code/find_check.rb
# frozen_string_literal: true

class FindCheck
  def initialize(board)
    @board = board
  end

  def find_king(board, piece)
    board.each_with_index do |row, row_index|
      col_index = row.index(piece)
      return row_index, col_index if col_index
    end
    nil
  end

  def piece_color(piece)
    piece == piece.upcase ? :white : :black
  end

  def pawn_check?(_pawn, row_index, col_index, king)
    potential_checks = [
      [row_index - 1, col_index - 1],
      [row_index - 1, col_index + 1]
    ]
    return true if potential_checks.include?(king)
  end

  def knight_check?(_piece, row_index, col_index, king)
    # Check if the king is in a knight's move away from the piece
    potential_checks = [
      [row_index - 2, col_index - 1],
      [row_index - 2, col_index + 1],
      [row_index - 1, col_index - 2],
      [row_index - 1, col_index + 2],
      [row_index + 1, col_index - 2],
      [row_index + 1, col_index + 2],
      [row_index + 2, col_index - 1],
      [row_index + 2, col_index + 1]
    ]
    return true if potential_checks.include?(king)

    false
  end

  def diagonal_check?(_piece, row_index, col_index, king)
    # Check if the king is on a diagonal path from the piece
    king_row, king_col = king
    return false unless (row_index - king_row).abs == (col_index - king_col).abs

    # Check if the path is clear of any other pieces
    x_dir = king_col > col_index ? 1 : -1
    y_dir = king_row > row_index ? 1 : -1
    x = col_index + x_dir
    y = row_index + y_dir
    while x != king_col
      return false if @board[y][x] != '0'

      x += x_dir
      y += y_dir
    end

    true
  end

  def vertical_check?(_piece, row_index, col_index, king)
    # Check if the king is on a vertical path from the piece
    king_row, king_col = king
    return false unless col_index == king_col

    # Check if the path is clear of any other pieces
    y_dir = king_row > row_index ? 1 : -1
    y = row_index + y_dir
    while y != king_row
      return false if @board[y][col_index] != '0'

      y += y_dir
    end

    true
  end

  def horizontal_check?(_piece, row_index, col_index, king)
    # Check if the king is on a horizontal path from the piece
    king_row, king_col = king
    return false unless row_index == king_row

    # Check if the path is clear of any other pieces
    x_dir = king_col > col_index ? 1 : -1
    x = col_index + x_dir
    while x != king_col
      return false if @board[row_index][x] != '0'

      x += x_dir
    end

    true
  end

  def check?
    white_king = find_king(@board, 'R')
    black_king = find_king(@board, 'r')
    @board.each_with_index do |row, row_index|
      row.each_char.with_index do |piece, col_index|
        king = piece_color(piece) == :white ? black_king : white_king
        next if %w[0 R r].include?(piece)

        case piece
        when 'P', 'p'
          return true if pawn_check?(piece, row_index, col_index, king)
        when 'A', 'a'
          return true if diagonal_check?(piece, row_index, col_index, king)
        when 'D', 'd'
          if diagonal_check?(piece, row_index, col_index, king) ||
             vertical_check?(piece, row_index, col_index, king) ||
             horizontal_check?(piece, row_index, col_index, king)
            return true
          end
        when 'T', 't'
          if vertical_check?(piece, row_index, col_index, king) ||
             horizontal_check?(piece, row_index, col_index, king)
            return true
          end
        when 'C', 'c'
          return true if knight_check?(piece, row_index, col_index, king)
        end
      end
    end
    false
  end
end
